doc/grabmode.man
.TH grabmode 8 "October 1998" "Version 1.9"

.SH NAME
.nf
grabmode \- a Textmode/XWindows mode grabber
clockprobe \- a Textmode/XWindows pixel clock probe
.fi

.SH SYNOPSIS
.B grabmode
[ -ndrhXTbcpv ] [ -x <width> ] [ -y <height> ]

.PP
.B clockprobe
[ -ndrhv ] [ -x <width> ] [ -y <height> ]

.SH DESCRIPTION
\fBclockprobe\fP tries to measure the pixel clock which is currently used
for the display that is visible while the probe is running, be it a textmode
display or a graphical one. Its default behaviour is to display the pixel
clock value in MegaHerz, together with the horizontal and vertical refresh
frequencies.

\fBGrabmode\fP basically does the same, but also grabs the entire state of
the VGA controller, and shows it on the screen in a format compatible with
the \fBXF86Config\fP file, the XFree86 X-server configuration file, or
alternatively in the \fBTextConfig\fP file format, which is the config file
format used by SVGATextMode.

Both programs are in fact the same, but depending on the name with which it
is called, it does a clockprobe or a grabmode (clockprobe is a link to
grabmode).

It does not change anything to the mode currently being probed. VGA
registers are read only. This means it does NOT actually change pixel clocks
like the X-server does when it probes for the clocks. This also implies that
it should work on ANY VGA card, even if it is unsupported by either XFree86
or SVGATextMode. This is because it only uses standard VGA registers. No
extended registers are used.

That also uncovers the biggest problem with grabmode: since it does not know
about extended registers, it doesn't know about special chipset-dependent
modes like 16/24/32-bit mode, or interlaced modes. See the \fBBUGS\fP
section for more details. The mode grabber tries to be as clever as possible
about interpreting the data retrieved from the VGA card, but it doesn't
always get it right. Using grabmode requires some common sense from the user
to interpret the data.

.SS Applications
Grabbing a standard VGA graphics mode (from svgalib or dosemu, or even from
DOS itself, using \fBgrabmode.exe\fP ) for inclusion in your own TextConfig
(or XF86Config) file is probably the most attractive one. "vgaset", included
in most X-distributions, can give you that, but it does NOT tell you the
pixel clock currently used, neither does it tell you about doublescan modes,
sync polarities, etc. The output of grabmode can be pasted right into your
XF86Config file (if it's correct...).

Of course, you could also grab the default Linux (and/or DOS) text modes to
include them in your TextConfig file. So you can show your friends how the
old modes looked, and how the SVGATextMode-"enhanced" modes look in contrast ;-)

Debugging XFree86 or SVGATextMode setups is another one. E.g. if
SVGATextMode or XFree86 fails to show you the kind of screen you'd expect,
and/or your monitor doesn't seem to like what you send it, then the most
common reason for that is that the pixel clock is not programmed correctly,
or not at all. Clockprobe or grabmode (even when typed blindly, redirected
to a file for later viewing) should be able to tell you what clock you are
ACTUALLY using. It's not because your "clocks" line says "50 MHz" for some
clock, that that clock index actually IS 50 MHz. You could be using the
wrong clocks line, or there could be a bug in the clock setting code. In
either case, the clockprobe can tell you the ACTUAL clock.

.SH OPTIONS
The first group of options are valid for both grabmode and clockprobe.

.TP
.B -n
don't touch VGA registers, not even to read them. Of course, nothing useful
will be done either. This option is more for completeness, and for debugging.

.TP
.B -d
debugging mode. grabmode/clockprobe will output lots of debugging
information about all the things it attempts to do. This is mostly for
"internal purposes" only, but it could help you discover why grabmode
doesn't react as you expect it to. This could also be useful when reporting
a problem.

.TP
.B -h
prints out a short help and the version number.

.TP
.B -r
"raw" mode. All smartwork will be skipped. The program will not attempt to
be smart about displaying the mode line and/or the clock calue. Grabmode
prints values as they are in the VGA registers, clockprobe shows the actual
clock used inside the VGA, not the per-pixel clock. See the \fPSMART BOMB\fP
section below for more information on how smart/stupid grabmode can be.

.TP
.B -v
verbose mode. Report all special mode flags. Only useful to VGA gurus.

.TP
.B -x <\fIwidth\fP>
.B -y <\fIheight\fP>
Specify width of the screen mode to be grabbed. In most cases, the screen
width and height are known to you in advance. E.g. you know you'll be
grabbing a 1024x768 mode, and all you want are the exact timings.

Specifying the width and/or height for the mode grabber or the clock probe
will bypass all the guessing (see "bugs" and "smart bombs" section) it would
normally have to do in order to detect special modes (15/16/24 BPP,
interlace, ...). This will result in a much more reliable mode line and/or
clock frequency report.

.PP
This option is for the clockprobe only:

.TP
.B -p
print only the clock value (in MHz), without ANY further comment. This is
probably useful in scripts and pipes.

.PP
These options are for grabmode only:

.TP
.B -X
grabmode normally checks to see if you are grabbing a text mode or a
graphics mode. It then adapts its output so it complies with a TextConfig
mode line (for text modes) or with an XF86Config mode line (for graphics
modes). This option forces it to output an XF86Config compatible mode line.

.TP
.B -T
The opposite of the -X option above: output a TextConfig mode line, even if
you're grabbing a graphics mode.

Since graphical screens have nothing to do with hardware assisted font
rendering (as from VGA text mode - let's forget about accelerators for a
while), they set the "character height" to 1 pixel (in fact they have to,
otherwise each line gets repeated 16 times, if they leave the "16" from the
9x16 character mode in there!). So font size given will of course be wrong.
In most cases you will be presented with am 8x1 font (8x2 in some DoubleScan
modes), and with a mode label equally distorted. grabbing a graphical
1024x768 mode this way would result in a mode line labelled "128x768" with a
font size of 8x1. You'll have to change the font size (and the label) if you
want to turn that into a useful text mode.

This could be used to create a text mode completely compatible with an
X-Windows mode you already have. switching from one to the other would be
much faster, since the monitor wouldn't have to re-synchronize.

.TP
.B -b
show video blanking information in addition to the normal mode line timing
information. Not very useful, unless you want to know everything about a VGA
mode. XF86Config nor TextConfig allow you to specify the blanking. Both
automatically program a suitable (though different) blanking. If you are
using the VGA overscan color for some reason, then this option will tell you
how far the overscan color will reach.

.TP
.B -c
Do not probe for the pixel clock, just print the mode line with the clock
replaced by "***". Not probing for the pixel clock makes grabmode a lot
faster.

.SH SAMPLE OUTPUT
.SS clockprobe
.nf
clockprobe: Estimated vertical scanrate = 69.754 Hz.
clockprobe: Estimated horizontal scanrate = 55.803 kHz.
clockprobe: Estimated pixel clock = 74.83 MHz
.fi

Note that the clockprobe (both in clockprobe and grabmode) only measures the
vertical refresh rate. The rest is calculated from the VGA register contents
and the vertical refresh. Extended and special modes which could cause
grabmode to mis-detect a mode, will also cause the horizontal frequency
and/or the pixel clock to be wrong.

.SS grabmode

(the output comes on one line)

.IP
"116x48"   74.832   928 968 1104 1192   768 775 776 800   +Hsync +Vsync  font 9x16   # 55.803kHz/69.75Hz

.SS grabmode -v
adds one more line to that:

.IP
#   Active special VGA register flags: WORDMODE

.SH SMART BOMBS
Grabmode, and to a lesser extent, clockprobe, have some intelligence
built-in, so they can cope with non-standard VGA configurations. As said
before, both programs use standard VGA registers only to get their
information. Only, nobody has a standard VGA card nowadays.

To remain chipset-independent, grabmode and clockprobe try to guess from
what they can read in the standard VGA registers what mode you could
actually be using.

A program can only be as intelligent as its creator (except by mistake).
Well, grabmode is extremely stupid on some occasions. Because of that, it
will report all non-trivial changes that it made to the original timings (as
read from the VGA) before printing it out, and also WHY it did that.

That's why this section is called "smart bombs": When one of those gets
confused...

The built-in smart (hah!) features are:

.SS DoubleScan detection.
This is a rather trivial feature, and is (for as far as I know) reliable
enough to go without warning. Doublescan is a way to display modes with very
few lines on a screen that was made to display many more. The most common
application is the CGA-emulation mode 320x200. Displaying only 200 lines on
a VGA screen would look awful (you'd see large black lines in between the
200 video lines).

Doublescan displays each line twice, so there are actually 400 lines on
screen, but only 200 different ones. In this mode, the VGA is programmed for
400 lines, but with Doublescan enabled, the ACTUAL line count is half that.
Grabmode will thus divide all vertical timings by 2 before displaying them.

.SS vertical timing wrap detection
The standard VGA register set only allows for a maximum of 2048 pixels
horizontally and 1024 lines vertically (this is the FULL size, including
sync and blanking, so the actual visible size is even smaller). Modern SVGA
cards can do modes with 1024 active lines and more, requiring more than 1024
lines in total. A 1280x1024 mode for example needs 1024 active lines. Add
sync and blanking time to that, and the grand total is over 1024 lines.

Although standard VGA has a "vertical double" mode bit, which multiplies all
vertical values by 2, most SVGA drivers prefer to use extended (chipset
dependent) registers to accomplish that. Grabmode cannot read these extended
registers (we want to keep it chipset-independent, remember?), so it will
show wrapped-around timings for those modes with around 1024 lines.

example:

.IP
Depending on how exactly a mode is implemented on your specific VGA card, a
1280x1024 mode at 60 Hz, which should be detected as:
.IP
.nf
"1280x1024"  110.34  1280 1328 1512 1712  1024 1025 1028 1054
.fi
.IP
will be detected (in "raw" mode) as:
.IP
.nf
"1280x1024"  3.131  1280 1328 1512 1712  1024 1 4 30
.fi

.PP
Grabmode will see that the sequence of vertical timings is wrong, and will
assume that this was an extended mode, using extended registers. It will
correct the sequence to make it look like the correct one.

This feature only works when the vertical timings happen to cross over the
10-bit standard VGA limit, which means it will only work for 1280x1024
modes. Far from perfect, but better than not at all...

.SS Golden ratio assumption
The "golden ratio" rule suggests that most displays have a physical aspect
ratio of 4x3 (width/height = 4/3). Most applications assume pixels are
square (same width and height), so the normal case would be to use the same
ratio in programming the VGA card.

640x480 modes have an aspect ratio of 640/480 = 1.333 (=4/3). 800x600,
1024x768, 1280x1024, 1600x1200: all those modes use an aspect ratio of 4/3
(or a close match) (anyone have a 16/9 computer display -- they exist) ?

If grabmode detects a mode that looks like double or triple that ratio, it
assumes this is a special mode (see below), and divides the horizontal
timings by 2 or 3. In most common cases, this is the correct result. An
appropriate warning is given.

The golden ratio assumption is then used by the CGA emulation detector, the
Hi-TrueColor mode detector and the interlace detector to adjust the mode
timings.

Grabmode (and clockprobe) report the _actual_ pixel clock used at which data
is being transferred from the VGA chip to the RAMDAC. This means that any
special data transfer scheme will mislead the clock probe, causing a faulty
report.

.SS Hi/TrueColor detection
HighColor and TrueColor modes on most cheap cards are one such example: the
VGA registers for e.g. a 640x480x16M mode on a normal VGA card with a
standard 8-bit RAMDAC (like most cheap VGA cards, e.g. using the S3 GenDAC
or ATT20C490/1/2) are set up as if the mode was actually 2x (16 bpp) or 3x
(24 bpp) or 4x (32 bpp) the actual horizontal width, and the pixel clock is
multiplied by the same value. So in that case a 24-bit 640x480 mode at 25
MHz pixel clock is programmed in the VGA chip as a 1920x480 mode at 75 MHz!
For this mode, the RAMDAC is programmed to assemble three consecutive bytes
into one giant 24-bit color pixel.

This ramdac programming is very RAMDAC-dependent, so it is impossible to
detect what is going on without knowing the RAMDAC type. RAMDAC's with
16-bit interfaces (like the S3 SDAC) use this scheme for 24/32-bit modes:
they transfer one 24-bit pixel in 2 clock periods, or, worse even, for
24-bit packed pixel modes, in 3/2 clock periods...

To make things even worse, grabmode will not even correctly report simple
256-color (8-bit) modes on some cards. Especially high-end cards (like #9,
Diamond, ...) using special VGA DAC's (Bt485,...) with pixel multiplexing.
They will report completely wrong clocks and refresh frequencies things on
higher (>67.5 MHz, for example) pixel clocks. Such RAMDAC's can transfer 2
or more entire pixels in one clock period, the opposite of simple RAMDAC's
in 16/24/32-bit modes.

.SS Interlace detection
On many (if not all) VGA cards, interlaced modes can be used to allow a high
resolution mode to be displayed on a low-end display (= with a low maximum
horizontal scan rate) without having to resort to ridiculously low vertical
refresh rates.

In order to do this, they program the VGA chip with HALF the amount of
actual lines on the screen, and set a special bit that tells the chip this
is an interlaced mode.

This "interlace-bit" is, again, in the extended (non-standard) VGA
registers, and can thus not be detected by the mode grabber. What _can_ be
detected is that the amount of lines programmed in the VGA registers is only
half of the height of the screen. This would be easy if there was a way to
know how much lines the screen actually has...

But since there isn't, grabmode must resort to some assumptions: assuming that
most graphics modes use one the standard screen sizes below:

.IP
.nf
640x480
800x600
1024x768
1152x900
1152x910
1280x1024
1600x1200
.fi

.P
the mode grabber can then check if the current mode has exactly HALF the
amount of lines of one of these modes: 240, 300, ..., 600 lines. If so, it
ASSUMES (!) that this is an interlaced mode, and changes the timings
accoringly before displaying them.

This immediately shows that the mode grabber will not be able to detect
non-standard interlaced modes.

.SS CGA emulation detection
A very popular graphics mode is the 320x200 mode with 256 colors. This is
CGA resolution, but with 256 colors (BIOS mode 13h): the only standard VGA
256 color mode. In this mode, the VGA reads 4 bits per pixel clock,
assembles each 2 groups of 4 bits into one 8-bit pixel and sends this to the
VGA. The actual VGA registers are thus programmed for a 640x400 mode: 400
lines with DoubleScan is 200 real lines, and 640 pixels with each 2 pixels
output as one is 320 real pixels. Why make it easy?

This pixel joining scheme is enabled by a certain bit in the VGA registers,
which is detected by grabmode to trigger it to transform the mode into a
"CGA" emulated mode. Easy.

BUT. Extended VGA modes use extended VGA registers, and those have priority
over the standard VGA registers. Hence, even if this special CGA-emulation
bit is set ON, it doesn't necessarily mean pixels are being combined.

All in all, even if this special mode is set in the standard VGA registers,
the extended registers could override that. Detecting this mode is thus not
trivial. This is why the "golden ratio" detection is used for this: after
doublescan has been accounted for, the mode timings come out as 640x200,
which triggers the golden ratio detector to transform this into 320x200,
with an appropriate warning.

.SS Vertical retrace timeout detection
The clock probe measures the average time between two vertical retraces
(busy-waiting. Yuck!). This means it will wait forever when there is no
vertical retrace signal. This can be caused either by a program actually
wrecking the current video mode (SVGATextMode has been known to do that :-)
or when the display system is in power-saving mode. 

When there is no vertical retrace detected within a few seconds, the probe
assumes it will never get one, and bail out with a message to that extent.
This avoids that the probe hangs forever.

.SH grabmode.exe & scanmode.exe (DOS port of grabmode)
There is a DOS port of grabmode in the SVGATextMode distribution, which
allows grabbing modes from dos.

There's one caveat to this. Previous versions of the DOS grabmode were
compiled using a commercial DOS C compiler, and required a lot of special
code. The latest version was compiled using DJGPP version 2.0, the GNU C
compiler for DOS. This required almost no special code, and thus it is a
cleaner port. BUT... It needs DPMI (DOS Proteced Mode Interface). A DMPI
environment can be found in MS-Windows (DOS boxes), OS/2, Windows-NT, qdpmi,
386max, etc. A free DMPI program is included in the distribution (The DJGPP
DPMI program CWSDPMI).

Scanmode.exe scans all VGA modes from the DOS BIOS, including the VESA
modes, and outputs two files in the currect directory: tconfig and xconfig.
The text modes are stored in the tconfig file, and the graphics modes are
stored in xconfig.

.TP
\fBIMPORTANT NOTE\fP:
The DOS port of all SVGATextMode programs is UNSUPPORTED. This means I am
mostly unable and, together with the people who helped me with this, rather
unwilling to put much effort in improving the DOS side of these tools.
Consider the DOS port a proof-of-concept. It was just plain fun to see the
DOS GCC compiler was so incredibly compatible with it's UNIX parent.

.SH BUGS
Is there an entry in the Guinness book of Records for
amount-of-bugs/code-size ratio? This program would come in second (just
behind MicroSoft).

.SS clock probing accuracy
The clock probe (used by clockprobe and grabmode) works best when the system
is not severely loaded. Especially disk access seems to have a bad
influence. DOSEMU is even worse. Very slow machines (i386) might run out of
steam while measuring. On 486 machines, the probe works fine.

However, even busy machines should give a reasonable result, unlike the X
clock probe. Even heavily loaded (swapping!) machines should give a rather
close estimate (maybe a few MHz off, but that should be all). System loads
to up to 3 or 4 will probably still yield results which are only a few MHz
wrong.

The clockprobe first performs 100 measurements of the time between two
vertical retraces, then calculates a histogram of those measurements. If the
histogram does not have a nice peak (mostly due to heavy system load,
causing many measurements to be wrong), the measurement is done all over
again. After three failures, the clock probe will stop retrying, using the
last measurement, but also warning you about the unreliability of the result.

In most cases, the clock probe will measure a value slightly higher or
slightly lower than you expected. It might be for example that you
programmed a mode line that specified a 70.000 MHz pixel clock, but the
clock probe insists that it is 75.175 MHz. This is due to the fact that any
clock generator cannot produce just _any_ clock you ask for, but must pick
the closest available one, either from a list (the \`Clocks\' line in
SVGATextMode or XFree86), or from the possible settings on a programmable
clock chip.

If the clockprobe insists on the same value over and over, it means that
that is the actual clock. If it insists on something far from the one you
expected, then it's either a bug in the probe, or a misconception on your
side.

Some special modes like CGA emulation, HiCilor, TrueColor, pixel
multiplexing, ..., will cause grabmode to fail to detect the correct number
of pixels per line. And because the horizontal sync and clock speed are all
calculated from the H and V timings, they too are wrong. The vertical sync
speed is ALWAYS correct, since it is MEASURED, independantly from ANY timing
register.


HISTORY
?? ??? ???? - version 1.9

- (1.8.1) remove "strip" option for "install" target (causes errors on older
  systems)
- fix sync polarity bug in Ti3026 Sync-On-Green code (Giacomo COMES)
- update vgaset's licensing to GPL (Igor Grobman/Debian, Greg Lehey)
- fix the bogus H/V VGA register upper limits. Especially vertically the
  limit was 4096, but should have been 1023 instead, which caused a lot of
  modes that appeared to stretch far below the bottom of the screen). Also
  fix all modelines that violated this constraint in the TextConfig file.
  (with help from Udo van den Heuvel)
- (1.8.2) some changes to STM_reset script (Suggested by Igor Grobman, Debian)
- DOS compile fixes and enhancements by Molnar Laszlo. New version of cwsdmpi
  included.
- don't bail out when ResetProg or FontProg doesn't exist (suggested by Igor
  Grobman, Debian)
- some FAQ additions
- fix S3 Sync-on-Green code for Ti3026. (Timo Giesselmann)
- don't allow ANYONE to run STM from a non-virtual terminal, add "-x" option
  to override this (suggested by Yann Dirson, Debian)
- set default fontpath to empty string, because "setfont" knows it as well
  (suggested by Yann Dirson, Debian)
- add small text explaining how to install STM using the SYSV init method in
  contrib/sysvinit (Michael De La Rue)
- add 40x25 mode (Otto J. Makela, Riku Saikkonen)
- (1.8.3) add Neomagic support (Pavel Machek)
- add "-k" option to disable software resize (hardware is resized) (Pavel
  Machek)
- (1.8.4) add MACH64 Rage II support (should now support all MACH64 boards
  with internal programmable clock)
- fix ET6000 MClk setting (use safer method recommended by Tseng)
- LOTS of work to program MACH64 DSP properly for textmode clocks. Thanks to
  Marc La France (of XFree86 fame) for inputs on this. 
- (1.8.5) integrate RIVA128 support from Martijn van Oosterhout.
- (1.8.6) fix bug in Neomagic code (option topleft didn't work)
- fix segfault in ET6000 "special" code
- force correct clockchip for some chips if not defined
- fix RIVA128 clock code (Martijn van Oosterhout)
- LOTS of extra free fonts (Frank Klemm)
- (1.8.7) Yikes... the RIVA code used a FIXED (#define) PCI base address.
  use PCI code now.


02 DEC 1997 - version 1.8

- add Matrox Millennium II support (jurriaan <thunder7@xs4all.nl>)


23 NOV 1997 - version 1.7

- (1.6.1) use /dev/tty0 instead of /dev/console (finally!)
- fix two minor bugs (missing break in vga_prg.c, new include for libc6)
  (Molnar Laszlo)
- added some more cards to PCI detected devices (avoids "PCI unknown device"
  message). Some cleanups in the PCI debug messages (was too verbose).
- Andree Borrmann's DOS/DJGPP patches (now Mystique should work in DOS)
- documentation updates (SIGWINCH, GPM, SDRAM, how to raise textmode max
  clock)
- (1.6.2) protect users from running STM from inside X or remotely (only
  root can do that now). (Benson L Chow)
- add "hshift" parameter to modelines (Massimiliano Ghilardi)
- (1.6.3) add support for Mystique rev 3 (Martin Scherbaum, S.u.S.e.)
- change setpalette-to-kernel color ordering from straight VGA to ANSI. This
  should make setpalette store the correct colors now. (Massimiliano
  Ghilardi)
- some Alpha compilation fixes (Antti Barck)
- (1.6.4) add Cirrus Laguna clockchip support (CLGD546x)
- add ISO_FONT9 option (Tuukka Toivonen)
- partial fix for IBM RGB RAMDAC problem with standard VGA clocks
- change "PROBLEMS" file into a FAQ. Maybe people will read it then...
- some DJGPP/DOS compilation fixes (missing #includes).


28 JUN 1997 - version 1.6

- (1.5.1) fixed bug in 9-pixel char mode (Stuart Lamble)
- fixed SIGSEGV in Cirrus clock code (Stanislav Voronyi)
- fixed bug in ET4000 W32i special_svga.c (linear address map comparator)
- (1.5.2) one more Tseng register is reset in special_svga.c (CRTC 0x33)
- fix bug in setVGAreg for 'io' registers (missing "break")
- add MATROX Mystique support (Trey Boudreau)
- (1.5.3) add a few new fonts (Massimiliano Ghilardi/Michele Maltoni)
- add DOS screen resizing patches and font loading from Andree Borrmann
- add video modelines text in contrib (PAL/NTSC)
- better DOS manual file formatting by jkekoni@aurora.tky.hut.fi
- MClk max limit is now 100 MHz instead of 70 MHz.
- the LSM file in version 1.5 was 1 year off :-)


15 APR 1997 - version 1.5

- (1.4.1) added "-k" option to setpalette (as requested)
- explicitly set a few more registers on ET4000/ET6000 to get better
  textmode restoration after X-server crash. Works much better now.
- SVGATextMode.c: added "-o" option, for "set80" functionality.
  (setting all std. VGA regs by force).
- setVGAreg.c: added "-c" and "-s" options for defining a set-mask or a
  clear-mask instead of an absolute value.
- DOS programs now use DJGPP cross-compiler from Linux.
  (doesn't work yet...)
- setVGAreg: "idx" and "io" 'registers' added for generic IO access.
- updated some docs (there was some obsolete stuff)
- fix bug in resize_specified_vts() (endless loop resizing first specified
  VT)
- only vga_prg.h includes asm/io.h now
- (1.4.2) fixed dos-gcc stuff in makefile: now "make dos" makes DOS
  executables out of the box (finally -- no more rebooting or dosemu)
- some stuff added to allow grabmode (not the others!) to compile under 
  the Cygnus GCC for Windows. Clock probing does not work yet.
- user_time() code added from Harald Koenig.
- (1.4.3) added Trident TGUI support contributed by Massimiliano Ghilardi.
- Integration of XFree 3.2 clock code (XFree/common_hw).
  THIS MAY CAUSE SOME SERIOUS NEW BUGS!
- (1.4.4) Mach64 support -- not finished yet. Not at all...
- 9-pixel wide characters now correctly display "box" characters (Stuart
  Lamble)
- added PCI bus code from XFREE/common_hw/xf86_PCI.* This will be needed for
  some of those PCI-only cards (Matrox Millennium et al).
- (1.4.6) Matrox Millennium added (not working yet)
- (1.4.7) ET4000 textmode restore enhancements
- Matrox Millennium code fixed -- ChipSet MATROX / ClockChip ti3026 (finished

grabmode.c
/*  SVGATextMode -- An SVGA textmode manipulation/enhancement tool
 *
 *  Copyright (C) 1995-1998  Koen Gadeyne
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 */


/***
 *** grabmode/clockprobe
 ***
 *** This program outputs the current text or graphics mode, in an SVGATextMode (or XF86Config-like) compatible line
 *** Should work on ANY VGA card, since it only uses standard VGA registers
 ***
 ***/

#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <unistd.h>

#include "misc.h"
#include "messages.h"
#include "modedata.h"
#include "mode_msg.h"
#include "string_ops.h"
#include "cfg_data.h"

#ifdef DOS
#define get_VGA_io_perm(x)
#undef Renounce_SUID
#define Renounce_SUID
extern int optind;
#endif

#ifdef win
#  include "asm/io.h"
#endif

char *CommandName;
bool debug_messages=FALSE;

int STM_Options=0;  /* just to keep the compiler happy */

#define CLOCKPROBE  1
#define GRABMODE    2

void usage(int func)
{
     PMESSAGE(("version %s. (c) 1995-1998 Koen Gadeyne.\n Usage: %s [options] \n"\
     "  Options: -n  Don't program VGA hardware\n"\
     "           -d  print debugging information\n"\
     "           -r  `raw' mode: don't do any guesswork on special modes,\n"\
     "               just print the values as they appear in the VGA registers\n"\
     "           -v  verbose mode (report all special VGA flags for the mode)\n"\
     "           -x <screen width>\n"\
     "           -y <screen height>\n"\
     "              specify screen width and height in pixels when it is known\n"\
     "              beforehand. This reduces the probe's guesswork (and errors).\n"\
     "           -h  print usage information\n"\
     "%s"\
     "  NOTE: All non-standard VGA features could cause bogus results:\n"\
     "         * interlaced modes\n"\
     "         * modes with >= 1024 active lines (like 1280x1024 and up)\n"\
     "         * HiColor (15/16bpp) and TrueColor (24/32bpp) modes\n"\
     "         * modes on cards with special DAC's\n"\
     "        The program will try to detect this and compensate\n"\
     "        (but will issue a warning that the actual VGA timings were adjusted)\n",\
     VERSION, CommandName,\
     (func==GRABMODE) ? \
     "           -X  Output compatible with XF86Config mode line\n"\
     "               (NOTE: grabbed graphic modes are not always correct)\n"\
     "           -T  Output compatible with TextConfig mode line\n"\
     "           -b  Also print blanking information\n"\
     "           -c  Don't probe for pixel clock (just print timing information)\n"\
     :\
     "           -p  print ONLY clock value (for use in scripts, pipes, etc)\n" ));
}

/***********************************************************************************************************/
 
int main (int argc, char* argv[])
{
  bool program_hardware=TRUE;
  int Xmode=MODE_UNDEFINED;
  bool showblank=FALSE;
  bool probe_clock=TRUE;
  bool pipeable=FALSE;
  bool raw_mode=FALSE;
  bool verbose=FALSE;
  int i;
  char* commandfilename;
  int c;
  modestruct m;
  int initial_x=-1, initial_y=-1;
  int func;
  
  CommandName = argv[0];
  commandfilename = strrchr(CommandName, '/');
  if (commandfilename) commandfilename++;
  else commandfilename = CommandName;

#ifdef DOS
  func = (!strncmp(commandfilename,"grab",4)) ? GRABMODE : CLOCKPROBE;
#else
  func = (!strncasecmp(commandfilename,"grab",4)) ? GRABMODE : CLOCKPROBE;
#endif

 
 /*
  * command-line argument parsing
  */

#ifdef WIN
   verbose=TRUE;
   debug_messages=TRUE;
#else
  while ((c = getopt (argc, argv, "ndrhXTbcpvx:y:")) != EOF)
    switch (c)
    {
      case 'n': program_hardware=FALSE;
                break;
      case 'd': debug_messages=TRUE;
                break;
      case 'r': raw_mode=TRUE;
                break;
      case 'X': Xmode = MODE_GRAPHICS;
                break;
      case 'T': Xmode = MODE_TEXT;
                break;
      case 'b': showblank = TRUE;
                break;
      case 'c': probe_clock = FALSE;
                break;
      case 'p': pipeable = TRUE;
                break;
      case 'v': verbose = TRUE;
                break;
      case 'h': usage(func);
                exit(0);
                break;
      case 'x': initial_x=getint(optarg, "X-size preset", 0, 4096);
                PDEBUG(("X-size on command line = %d\n", initial_x));
                break;
      case 'y': initial_y=getint(optarg, "Y-size preset", 0, 4096);
                PDEBUG(("Y-size on command line = %d\n", initial_y));
                break;
      case '?': usage(func);
#ifndef DOS
                PERROR(("Bad option '-%c'\n",(char)optopt));
#endif                
                exit(-1);
                break;
      default: PERROR(("getopt returned unknown token '%c'.\n",c));
    }
#endif

 PVERSION;

/*
 * start doing something useful
 */
 
 if (!program_hardware) return(0);

 PWARNING(("Please be patient. This may take a while (up to 1 minute)\n"));

 getmode(&m, probe_clock, raw_mode, initial_x, initial_y);

 if (func == CLOCKPROBE)
 {
   if (pipeable)
   {
     printf("%1.2f\n", m.mode_line.pixelClock/1000.0);
   }
   else
   {
     PMESSAGE(("Estimated vertical scanrate = %1.3f Hz.\n", m.mode_line.vfreq/1000.0));
     PMESSAGE(("Estimated horizontal scanrate = %1.3f kHz.\n", m.mode_line.hfreq/1000.0));
     PMESSAGE(("Estimated pixel clock = %1.2f MHz\n", m.mode_line.pixelClock/1000.0));
     if (GMOFLG_ISSET(m,CLOCKDIV2))
       PMESSAGE(("NOTE: actual clock is %1.2f MHz, but it is divided by 2.\n", m.mode_line.pixelClock*2/1000.0));
   }
   return(0);
 }

 /*
  * "grabmode" function.
  * From parameters, look for special settings that affect the output
  */
  
  /* auto selection between graphics/text mode if not defined on command line */
  if (Xmode==MODE_UNDEFINED) Xmode = m.txt_gr_mode;
  
 /*
  * Print the mode line
  */
  
 if (Xmode==MODE_TEXT)
   printf("\"%dx%dx%d\"   ", m.mode_line.HDisplay/8, m.mode_line.VDisplay/m.mode_line.FontHeight, m.mode_line.FontWidth);
 else 
   printf("\"%dx%d\"   ", m.mode_line.HDisplay, m.mode_line.VDisplay);

 if (probe_clock)
   printf("%1.3f", m.mode_line.pixelClock/1000.0);
 else
   printf("***");
         
 printf("   %d %d %d %d   %d %d %d %d   %cHsync %cVsync",
         m.mode_line.HDisplay, m.mode_line.HSyncStart, m.mode_line.HSyncEnd, m.mode_line.HTotal,
         m.mode_line.VDisplay, m.mode_line.VSyncStart, m.mode_line.VSyncEnd, m.mode_line.VTotal,
        (m.mode_line.hpol==POS) ? '+' : '-',
        (m.mode_line.vpol==POS) ? '+' : '-');
        
 if (GMOFLG_ISSET(m,DOUBLESCAN)) printf("  DoubleScan");
 if (GMOFLG_ISSET(m,MULTISCAN)) printf("  %sScan", m.mode_line.FontHeight==2 ? "Double" : "Multi");

 if (GMOFLG_ISSET(m,INTERLACE)) printf("  Interlace");

 if (m.mode_line.hshift) printf("  hshift %d", m.mode_line.hshift);

 if (Xmode==MODE_TEXT) printf("  font %dx%d", m.mode_line.FontWidth, m.mode_line.FontHeight);
 
 if (probe_clock)    
   printf("   # %1.3fkHz/%1.2fHz\n", m.mode_line.hfreq/1000.0, m.mode_line.vfreq/1000.0);
 else
   printf("\n");
   
 if (showblank)
   printf("#   Blanking: H=%d-%d ; V=%d-%d\n", m.starthbl, m.endhbl, m.startvbl, m.endvbl);
   
 if (verbose)
 {  
   printf("#   Active special VGA register flags: ");
   if (m.mode_flags == 0) printf("none.");
   if (GMOFLG_ISSET(m,DOUBLESCAN))      printf("DOUBLESCAN ");
   if (GMOFLG_ISSET(m,CLOCKDIV2))       printf("CLOCKDIV2 ");
   if (GMOFLG_ISSET(m,MULTISCAN))       printf("MULTISCAN ");
   if (GMOFLG_ISSET(m,BYTEMODE))        printf("BYTEMODE ");
   if (GMOFLG_ISSET(m,WORDMODE))        printf("WORDMODE ");
   if (GMOFLG_ISSET(m,DOUBLEWORDMODE))  printf("DOUBLEWORDMODE ");
   if (GMOFLG_ISSET(m,VERT_DOUBLE))     printf("VERT_DOUBLE ");
   if (GMOFLG_ISSET(m,PCS_DIV2))        printf("PCS_DIV2 ");
   if (GMOFLG_ISSET(m,CNT_BY_2))        printf("CNT_BY_2 ");
   if (GMOFLG_ISSET(m,INTERLACE))       printf("INTERLACE ");
   if (GMOFLG_ISSET(m,SHIFT_CGA256C))   printf("SHIFT_CGA256C ");
   if (GMOFLG_ISSET(m,SHIFT_CGA4C))     printf("SHIFT_CGA4C ");
   printf("\n");
 }  
 
 if (m.remarks!=0)
 {
   PDEBUG(("Remarks mask: 0x%4X\n", m.remarks));
   fputc('\n', stderr);
   PWARNING(("Grabbed mode could be wrong. Below is a list of remarks about it:\n\n"));
   for (i=0; mode_messages[i].msg_mask >0 ; i++)
   {
     if ((m.remarks & (1<<i)) != 0)
     {
       if ((1<<i)==MSG_CLOCK_MEASUREMENTS)
         fprintf(stderr, mode_messages[i].msg_txt,m.valid_measurements);
       else
         fprintf(stderr, mode_messages[i].msg_txt);
       fputc('\n', stderr);
     }
   }
   return(1);
 }
  return(0);
}
